Ruby bindings

BernhardDenner opened the pull request:
# Purpose

This is the initial Pull request for libelektra Ruby bindings implemented using SWIG:
- bindings for kdb::Key, kdb::KeySet and kdb::KDB
- exception handling implemented
- KeySet supports kdb iterator through current/next and Ruby style iterators (Enumerable mixin)
- follows Ruby naming conventions
- includes unit tests 
- includes examples

However, documentation could be better.
# Checklist
- [x] commit messages are fine (with references to issues)
- [x] I ran all tests and everything went fine
- [x] I added unit tests
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions
- [x] meta data is updated (e.g. README.md of plugins)

@markus2330 please review my pull request, mainly check the unit test cases, if it looks promising

markus2330 commented on src/bindings/swig/ruby/tests/testruby_keyset.rb:
```diff
@@ -48,4 +48,102 @@ def test_keySet_append
+      ks = Kdb::KeySet.new
+
+      # populate keySet
+      a.each { |e| ks << e }
```
Very elegant! What about testing different ways of keyset construction with always using the checks as below?

BernhardDenner replied:
the temp. Array 'a' was just used to allow native access to original Keys for testing the KeySet methods.
The '<<' operator is also implemented for the KeySet.
`ks = Kdb::KeySet.new
ks << Kdb.Key.new("user/key", value: "hello world")`

The va_list for the KeySet.new is on the TODO-List to allow something like
`ks = Kdb::KeySet.new [Kdb::Key.new("user/k1"), Kdb::Key.new("user/k2")]`

markus2330 replied:
> Array 'a' was just used to

Yes, I meant that something like this is needed often, so it is nice if it is easy to transform build-in data structures to KeySet and the other way round.

Similar unit tests exist for C++, too. There is also a unit-test that appends key in every possible order and checks if all of those keysets behave identical. It is not necessary to do this in ruby, too, but every constructor should be tested.

> `ks = Kdb::KeySet.new [Kdb::Key.new("user/k1"), Kdb::Key.new("user/k2")]`

Yes, that would be nice, too!

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -1,15 +1,33 @@
+ * method names are renamed to follow Ruby naming conventions
+ * Key and KeySet methods directly modify the underlaying Key/KeySet
+
+Please note, this documentaition will show C++ types also (e.g. std::string).
```
typo: documentaition

What means that it "will show C++ types also"?

markus2330 replied:
english: "C++ types, too" or "will also show"

markus2330 commented on src/bindings/swig/ruby/examples/ruby_example_application.rb:
```diff
@@ -0,0 +1,123 @@
+    # now we are ready to save the updated settings
+    ret = db.set ks, APP_NS
+rescue KDBException
+    puts "clould now update KDB database: %s" % $!
```
typo: could

markus2330 replied:
Not fixed.

markus2330 commented on src/bindings/swig/ruby/examples/ruby_example_application.rb:
```diff
@@ -0,0 +1,123 @@
+def pretty_print_key( key )
+    puts "key '%s'" % key.name
+    if key.is_binary?
+        puts "  binary key, lenght: %d" % key.value.length
```
typo

markus2330 replied:
still open

markus2330 commented on src/bindings/swig/ruby/examples/ruby_example_keyset.rb:
```diff
@@ -0,0 +1,143 @@
+ks_a = Kdb::KeySet.new Kdb::Key.new("user/key1")
+ks_b = Kdb::KeySet.new Kdb::Key.new("user/key1")
+
+ks_a == ks_b  # => true
```
Is there also an identity operation that says false?

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+
+%feature("autodoc", "3");
+
+%define CPPDOCURL "http://www.libelektra.org/TBD!!!" %enddef
```
fix TBD

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+ * Access to native C-level KDB structures (such as ckdb::Key) is not
+   possible, as this does not make much sense within Ruby.
+ * method names are renamed to follow Ruby naming conventions
+ * Key and KeySet methods directly modify the underlaying Key/KeySet
```
typo: underlying

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+%enddef
+/* docstring for module wasn't implemented until now, however, there's 
+   a pull-request (https://github.com/swig/swig/pull/582), thus this feature
+   should be available soon */
```
is already closed.

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+%predicate kdb::Key::isBelowOrSame;
+%predicate kdb::Key::isDirectBelow;
+%predicate kdb::Key::hasMeta;
+%predicate kdb::Key::isNull; // TODO: do we need something special here??? 
```
TODO?

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+%rename("name") kdb::Key::getName;
+%rename("name=") kdb::Key::setName;
+
+%rename("base_name") kdb::Key::getBaseName;
```
Maybe rename to "basename" (as in ruby's File API)?

BernhardDenner replied:
yes, good idea
did that also for `Key.add_basename`

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+
+/* 'imitate' va_list as Ruby Hash
+ * 
+ * "missuse" the exception feature of SWIG to provide a custom
```
typo: misuse

markus2330 replied:
Is the exception feature in ruby expensive? Can you run some small time measurement benchmarks?

BernhardDenner replied:
not more expensive the standard c++ exception handling.
The swig exception feature allows to replace the target function invocation to be replaced by your own code, in order to be able to place your custom exception handling mechanisms. Normally you would use something like
```
try {
    $action  // will be replace by SWIGs target function invocation
catch (...) {
    ...
}
```
However, I've used this feature to place my own target function call instead. In this case this will not much take longer than the standard way.
For details you can look into the SWIG generated C-file.

markus2330 replied:
That is why I asked: Exceptions in C++ cause no overhead, unless they are actually thrown; then they are expensive. But it might be irrelevant in relation to what ruby does.

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+  int i;
+  int flags = 0;
+
+  /* $input substitution does not here, so we have to reverence
```
reverence?

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+      val = rb_funcall(val, rb_intern("to_s"), 0, NULL);
+      /* ignore certain keys */
+      if (strcmp("flags", StringValueCStr(key)) == 0) continue;
+      /* 'value' has also a special meening */
```
typo meening

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+      val = rb_hash_aref(argv[1], key);
+      key = rb_funcall(key, rb_intern("to_s"), 0, NULL);
+      val = rb_funcall(val, rb_intern("to_s"), 0, NULL);
+      /* ignore certain keys */
```
Why only these and not others (see keyVInit in src/libs/elektra/keyhelpers.c)?

BernhardDenner replied:
actually, I've looked into this function and I think these two are the only reasonable here.
- `size` doesn't make sense, since Rubys `setBinary` doesn't take a size parameter (Ruby uses Strings for binary values and Strings have a `size` method).
- `func` is currently not implemented
- `meta`: everything except `value` or `flags` is considered as metadata
- other flags will be passed through `flags` key
- and the others (`KEY_NAME`, `KEY_OWNER`...) is considered as deprecated

markus2330 replied:
Can you add this to documentation please?

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+
+        rb_yield(cur);
+
+        /* TODO: do we have to free anything ? */
```
TODO

Did you try to check with valgrind? Currently there seems to be a bunch of memleaks.

BernhardDenner replied:
In this case, this isn't a memleak, since the new Ruby obj. will be the owner of the newly created key.
Latest valgrind runs didn't show any memleaks.

markus2330 commented on src/bindings/swig/ruby/kdb.i:
```diff
@@ -0,0 +1,691 @@
+    
+  } else {
+  /* standard case for KeySet, just convert and check for correct type */
+    //fprintf(stderr, "append KeySet\n");
```
Use logger?

markus2330 commented on src/bindings/swig/ruby/kdb.rb:
```diff
@@ -0,0 +1,74 @@
+# This is not required any more
```
Can be removed then.

markus2330 commented on src/bindings/swig/ruby/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+        get_filename_component (name ${file} NAME_WE)
+        add_test (
+                NAME ${name}
+                COMMAND ${RUBY_EXECUTABLE} ${file} --verbose=verbose 
```
verbose flag? see above

markus2330 commented on src/bindings/swig/ruby/tests/testruby_kdb.rb:
```diff
@@ -0,0 +1,86 @@
+
+      ret = h.get ks, RB_TEST_NS
+
+      assert_true ret >= 0
```
See above (assert_true, assert_false missing)

markus2330 commented on src/bindings/swig/ruby/tests/testruby_kdb.rb:
```diff
@@ -0,0 +1,86 @@
+
+      assert_equal 1, ret
+
+      h.close
```
Is it possible to pass parent key here? (For warnings)

markus2330 commented on src/bindings/swig/ruby/tests/testruby_kdb.rb:
```diff
@@ -0,0 +1,86 @@
+      assert_equal 1, ret 
+
+      # use KDB_O_POP to remove the keys from the KeySet
+      k = ks.lookup "#{RB_TEST_NS}/kdbgetset/c1", Kdb::KDB_O_POP 
```
Have a method `delete`, `delete_at` and so on to be more ruby like?

markus2330 commented on src/bindings/swig/ruby/tests/testruby_key.rb:
```diff
@@ -0,0 +1,390 @@
+      k = Kdb::Key.new name
+      assert_true k.is_valid?
+      assert_equal name, k.name
+      assert_equal "", k.get_string # TODO: should we return nil here
```
TODO

markus2330 commented on src/bindings/swig/ruby/tests/testruby_key.rb:
```diff
@@ -0,0 +1,390 @@
+
+      k.add_name "../../../../new1"
+      assert_equal "new1", k.base_name
+      assert_equal "#{name}/new1", k.full_name
```
tests with escaped slashes missing?

markus2330 commented on src/bindings/swig/ruby/tests/testruby_keyset.rb:
```diff
@@ -0,0 +1,615 @@
+      assert_equal "user/ks1", ks.head.name
+    end
+
+    assert_raise ArgumentError do
```
assert_raise more specifically (where it should happen)?

markus2330 commented on src/bindings/swig/ruby/tests/testruby_keyset.rb:
```diff
@@ -0,0 +1,615 @@
+      ks << Kdb::Key.new("user/k1")
+
+      assert_equal 1, ks.size
+      assert_equal 1, ks.length
```
some more appends and pop?

markus2330 commented on src/bindings/swig/ruby/tests/testruby_keyset.rb:
```diff
@@ -0,0 +1,615 @@
+    assert_nothing_raised do
+      ks = Kdb::KeySet.new (0..5).map { |i| Kdb::Key.new "user/key#{i}" }
+
+      a = ks.to_a
```
Nice!

markus2330 commented on src/bindings/swig/ruby/tests/testruby_keyset.rb:
```diff
@@ -0,0 +1,615 @@
+      # with options
+      #lookupkey = Kdb::Key.new "user/key04"
+      #assert_equal lookupkey, ks.lookup(lookupkey, Kdb::KDB_O_DEL)
+      #assert_true lookupkey.is_null? # TODO: should this really work this way
```
TODO

markus2330 commented on src/bindings/swig/ruby/tests/testruby_kdb.rb:
```diff
@@ -0,0 +1,86 @@
+      h.set ks, RB_TEST_NS
+    end
+  end
+end
```
compare with KDB_VERSION? (See python test cases)

markus2330 commented:
In 2e0ecf2d0d37a0c4c4cbf04654f002fd319aa943 I added `is*` of other namespaces, too. I think this requires changes in this PR (rename of methods).

markus2330 commented on src/bindings/swig/ruby/CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+
+# try user environment first, then search for system libs
+# required on OSX under rvm
+set (CMAKE_FIND_FRAMEWORK "LAST")
```
Can you make this for OSX only?

markus2330 commented on src/bindings/swig/ruby/CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+	include_directories(${RUBY_INCLUDE_DIRS})
+	
+	# just for debugging
+	message (STATUS "Ruby executable: ${RUBY_EXECUTABLE}")
```
comment out please

markus2330 commented on src/bindings/swig/ruby/examples/ruby_example_application.rb:
```diff
@@ -0,0 +1,123 @@
+# helper methods
+#
+
+def pretty_print_key( key )
```
any news here?

BernhardDenner commented:
Ok, seams that I have forgotten some things.
Now, I've
- fixed the remaining issues
- added `to_s` and `pretty_print` to `Key` and `KeySet`
- added a `open` method, similar to `File.open` semantic
- added documentation

markus2330 commented:
Thank you, great job!
